framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java

```diff
@@ -2154,24 +2041,115 @@ public boolean shouldSuppressWarnings(@Nullable Element elt, String errKey) {
+                Arrays.setAll(suppressWarningStrings, i -> suppressWarningStrings[i].toLowerCase());
+                if (shouldSuppressWarningStringSuppressKey(suppressWarningStrings, errKey)) {
+                    if (hasOption("warnUnneededSuppressions")) {
+                        elementsWithSuppressedWarnings.add(elt);
+                    }
+                    return true;
+                }
             }
-            return true;
+            if (isAnnotatedForThisCheckerOrUpstreamChecker(elt)) {
+                // Return false immediately. Do NOT check for AnnotatedFor in the
+                // enclosing elements, because they may not have an @AnnotatedFor.
+                return false;
+            }
+            elt = elt.getEnclosingElement();
         }
+        return false;
+    }
 
-        if (isAnnotatedForThisCheckerOrUpstreamChecker(elt)) {
-            // Return false immediately. Do NOT check for AnnotatedFor in the
-            // enclosing elements, because they may not have an @AnnotatedFor.
+    /**
+     * Determines whether an error (whose message key is {@code messageKey}) should be suppressed.
+     * It is suppressed if any of the given SuppressWarnings strings suppresses it.
+     *
+     * <p>A SuppressWarnings string may be of the following pattern:
+     *
+     * <ol>
+     *   <li>{@code "prefix"}, where prefix is a SuppressWarnings prefix, as specified by {@link
+     *       #getSuppressWarningsPrefixes()}. For example, {@code "nullness"} and {@code
+     *       "initialization"} for the Nullness Checker, {@code "regex"} for the Regex Checker.
+     *   <li>{@code "partial-message-key"}, where partial-message-key is a prefix or suffix of the
+     *       message key that it may suppress. So "generic.argument", would suppress any errors a
+     *       message key that contains to generic.argument.
+     *   <li>{@code "prefix:partial-message-key}, where the prefix and partial-message-key is as
+     *       above. So "nullness:generic.argument", would suppress any errors in the Nullness
+     *       Checker with a message key that contains to generic.argument.
+     * </ol>
+     *
+     * {@code "allcheckers"} is a prefix that suppresses a warning from any checker. {@code "all"}
+     * is a partial-message-key that suppresses a warning with any message key.
+     *
+     * <p>If {@code -ArequirePrefixInWarningSuppressions}, then {@code prefix} and {@code
+     * prefix:partial-message-key} are the only SuppressWarnings strings that have an effect.
+     *
+     * @param suppressWarningStrings the SuppressWarnings strings. May be null, in which case this
+     *     method returns false.
+     * @param messageKey the message key of the error the checker is emitting; a lowercase string,
+     *     without any "checkername:" prefix
+     * @return true if one of the {@code suppressWarningStrings} suppresses the error
+     */
+    private boolean shouldSuppressWarningStringSuppressKey(
```

Comment:
The implementations of
  `SourceChecker.shouldSuppressWarningStringSuppressKey`
  `SourceChecker.warnUnneededSuppressions`
are very similar.
Either abstract out common logic, or eliminate one of these two methods.
(Note that there are some undesirable differences, such as the logic for permitting a user-written messagekey to be a substring of a full Checker Framework messagekey.)